### PR TITLE
Start roadmap

### PR DESCRIPTION
**Pull request by raphlinus**
This is very much a half-finished draft, but I'm creating a PR for it now because it might be useful.

### CODE REVIEW

File: docs/roadmap.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+In addition, many of these crates are designed as common infrastructure, and could be used by other GUI efforts.
+
+### druid
+
+### druid-shell
+
+The current scope of druid-shell is similar to that of winit: it pops up a window, and takes inputs. There is one additional set of functionality: it gives access to platform menus.
+
+I didn't use winit because I wanted to experiment, and didn't know whether design decisions of winit would limit performance or functionality. Shortly I want to take a close look at those questions. I can imagine using winit for that common functionality and layering other functionality on top.
```
Comment (rylev) [inline docs/roadmap.md]: Some thoughts on what additional backends you'd like to support in the short and medium term would be nice

File: docs/roadmap.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+## Future: an higher level API surface for UI
+
+A `Widget` in druid is roughly comparable to a `RenderObject` in Flutter. I'm personally comfortable writing UI's directly at this level, but at the same time it might not be ideal. Flutter implements a reactive layer (what it calls Widgets) on top. In my opinion, the best way to do a high level API surface in Rust is an open research question. In additional to elm-style, I can imagine something that resembles imgui, something that expands templates, and something like Interface Builder. There are many interesting details about how state is managed, whether there's interior mutability, whether macros make the syntax sweet or whether it's possible to do something reasonably clean without relying heavily on macros.
+
+So basically I invite the community to explore this.
```
Comment (rylev) [inline docs/roadmap.md]: Does 'this' mean high level gui frameworks? Or all the things you laid out in the post? It's not super clear whether you intend to explore this space yourself...

File: docs/roadmap.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+I didn't use winit because I wanted to experiment, and didn't know whether design decisions of winit would limit performance or functionality. Shortly I want to take a close look at those questions. I can imagine using winit for that common functionality and layering other functionality on top.
+
+### piet and back-ends
+
+I explained the motivation of piet in a [blog post](https://raphlinus.github.io/rust/graphics/2018/10/11/2d-graphics.html). Since then, it's coming along pretty nicely. There are three back-ends: Direct2D which is high performance and Windows-only, Cairo which is mature and portable, but not cleanly packaged as Rust crates and not very performant, and web.
+
+### As yet unnamed high-level text crate
+
+The biggest missing functionality in piet is text handling. What's there right now is very primitive. I think DirectWrite is a good model for what apps need, but of course it's Windows specific.
```
Comment (rylev) [inline docs/roadmap.md]: I suppose you assume most people who care about this will know what directwrite does but writing them out might be nice...

**Comment by cmyr** [comment]
@raphlinus shall we put this out of its misery?

**Comment by raphlinus** [comment]
Yes. Definitely want to do an updated roadmap sometime in the next few weeks though.

**Pull request closed without merging**